lib/postgrex/protocol.ex

```diff
@@ -643,106 +645,90 @@ defmodule Postgrex.Protocol do
     # TODO: subscribers
     status
   end
-  defp handle_msg(status, msg_ready()) do
-    status
-  end
 
-  defp ok(_, status, buffer) do
-    %{parameters: parameters, notifications: notifications} = status
-    {:ok, parameters, Enum.reverse(notifications), buffer}
-  end
-
-  defp ok(result, _, status, buffer) do
-    %{parameters: parameters, notifications: notifications} = status
-    {:ok, result, parameters, Enum.reverse(notifications), buffer}
+  defp sync(s, status, result, buffer) do
+    case msg_send(msg_sync(), s) do
+      :ok ->
+        sync_recv(s, status, result, buffer)
+      {:error, _} = err ->
+        err
+    end
   end
 
-  defp error(reason, _) do
-    {:error, reason}
+  defp sync_recv(s, %{mode: mode} = status, result, buffer) do
+    %{sock: sock, timeout: timeout} = s
+    case msg_recv(sock, buffer, timeout) do
+      {:ok, msg_ready(), buffer} ->
+        %{parameters: parameters, notifications: notifications} = status
+        {:ok, result, parameters, Enum.reverse(notifications), buffer}
```

Comment:
In order to check only one command we need to receive the `msg_ready()` packet in the query. This required quite a bit of refactoring but actually ended up removing about 60 lines.

It would be interesting if we could make use of the status in the `msg_ready` packet (`:failed`, `:idle`, `:transaction`) to guarantee that a connection checked out from a pool that is `:idle`. Note that result here may not be `Postgrex.Result.t`, it can be `Postgrex.Error.t` or `{kind, reason, stack}` as there are many errors we can recover from.